### PR TITLE
Add samourai-server app

### LINKED ISSUE #458: Package Samourai Dojo
Prior art: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/docker-compose.yaml

I believe those containers are x86 only so we'll need some multi-arch containers.

Also there are some config/source files than need manaully populating with tokens.

Also some manual database migration logic that needs handling for updates.

### LINKED ISSUE #459: Package Samourai Whirlpool
Prior art: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/overrides/whirlpool.install.yaml

A whirlpool app could just be the Whirlpool CLI with the app UI simply showing a connection string: https://code.samourai.io/whirlpool/whirlpool-client-cli

The actual UI would be handled via the Samourai mobile app.

However ideally it would the app would bundle the Whirlpool GUI: https://code.samourai.io/whirlpool/whirlpool-gui

Whirlpool GUI is an Electron app, so I'm not sure how easily it can exposed as just a web app from Umbrel. However if it can be done without too much trouble this will give a far superior experience.

### PR DESCRIPTION
**Pull request by louneskmt**
# App Submission

### App name

Samourai Server

### Version

1.8.0

### One line description of the app
_(max 50 characters)_

Your private backing server for Samourai Wallet

### Summary of the app
_(50 to 200 words)_

Samourai Server is an exclusive Umbrel app that runs both the Samourai Dojo and Whirlpool servers, and provides you easy step-by-step instructions to connect your Samourai Wallet to them. Samourai Wallet is unrivaled in transaction privacy, but its default configuration is still subject to network level privacy loss.

By default, Samourai Wallet will connect to its centrally controlled Dojo hosted in Iceland. By hosting your own Samourai  Server, you can bypass its servers completely when using Samourai Wallet or Sentinel, and gain total financial sovereignty without any complexity.

### Developed by

Samourai 

### Developer website

https://samouraiwallet.com

### Source code repository

https://code.samourai.io/dojo/samourai-dojo

### Support link
_(Link to your Telegram support channel, GitHub issues/discussions, support portal, or any other place where users could contact you for support.)_

https://t.me/SamouraiWallet

### Uses
- [x] Bitcoin Core
- [x] Electrum server
- [ ] LND

### 256x256 SVG icon
_(GitHub doesn't allow uploadig SVGs directly. Upload your file to an alternate service, like https://svgur.com, and paste the link below.)_

icon

### App screenshots
_(Upload 3 to 5 high-quality screenshots (at least 1280x800px) of your app in PNG format.)_

...

### I have tested my app on:
- [ ] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [ ] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [ ] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

Closes #458
Closes #459

### CODE REVIEW

File: scripts/app [hunk 1]
```diff
@@ -88,6 +88,7 @@ compose() {
 
   export BITCOIN_DATA_DIR="${UMBREL_ROOT}/bitcoin"
   export LND_DATA_DIR="${UMBREL_ROOT}/lnd"
+  export TOR_DATA_DIR="${UMBREL_ROOT}/tor/data"
```
Comment (lukechilds) [inline scripts/app:91]: Why is this exported?

We shouldn't allow apps to mount the Tor data dir, even as read only, it contains the private keys for all Tor hidden services.

File: templates/bitcoin-sample.conf [hunk 2]
```diff
@@ -19,8 +19,9 @@ maxconnections=40
 maxuploadtarget=1000
 
 # zmq
-zmqpubrawblock=tcp://0.0.0.0:28332
-zmqpubrawtx=tcp://0.0.0.0:28333
+zmqpubrawblock=tcp://0.0.0.0:<port>
+zmqpubrawtx=tcp://0.0.0.0:<port>
+zmqpubhashblock=tcp://0.0.0.0:<port>
```
Comment (lukechilds) [inline templates/bitcoin-sample.conf:24]: It's fine to expose these options to apps but ideally it would be done in a separate PR.

**Review by lukechilds — CHANGES_REQUESTED** [review]
I wasn't able to run this locally due to IP collisions but here's some feedback based on skimming the code.

File: apps/samourai-dojo/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  
+  node:
+    image: "louneskmt/dojo-nodejs:1.8.0"
+    logging: *default-logging
+    restart: always
+    command: "/home/node/app/wait-for-it.sh db:3306 --timeout=720 --strict -- /home/node/app/restart.sh"
+    environment: 
+      # GLOBAL
+      COMMON_BTC_NETWORK: $BITCOIN_NETWORK
+      DOJO_NODEJS_VERSION_TAG: 1.8.0
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:33]: Needed here: https://code.samourai.io/dojo/samourai-dojo/-/blob/v1.8.1/docker/my-dojo/node/keys.index.js#L37

File: apps/samourai-dojo/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      BITCOIND_ZMQ_RAWTXS: $BITCOIN_ZMQ_RAWTX_PORT
+      BITCOIND_ZMQ_BLK_HASH: $BITCOIN_ZMQ_HASHBLOCK_PORT
+      BITCOIND_SHUTDOWN_DELAY: 180
+      BITCOIND_EPHEMERAL_HS: "off"
+
+      # EXPLORER  
+      EXPLORER_INSTALL: "off"
+
+      # INDEXER
+      INDEXER_INSTALL: "off"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:83]: `EXPLORER_INSTALL` needed here: https://code.samourai.io/dojo/samourai-dojo/-/blob/v1.8.1/docker/my-dojo/node/keys.index.js#L17

But you're right for `INDEXER_INSTALL`, I missed this one (I checked almost all env var to see where they are used).

File: scripts/app (hunk 1 above)
Comment (louneskmt) [reply scripts/app:91]: Following a discussion on Keybase, this is not needed and will be removed.

File: templates/bitcoin-sample.conf (hunk 2 above)
Comment (louneskmt) [reply templates/bitcoin-sample.conf:24]: Gonna let this here for testing purpose, remove it before merge and make another PR. Is that okay for you?

File: apps/samourai-dojo/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    stop_grace_period: 5m
+    environment:
+        MYSQL_DATABASE: samourai-main
+        MYSQL_ROOT_PASSWORD: rootpassword
+        MYSQL_USER: samourai
+        MYSQL_PASSWORD: password
+    volumes:
+      - ${APP_DATA_DIR}/mysql:/var/lib/mysql
+    expose:
+      - "3306"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:23]: Will be removed.

File: apps/samourai-dojo/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    image: "louneskmt/dojo-nginx:1.5.0"
+    logging: *default-logging
+    restart: always
+    command: "/wait-for node:8080 --timeout=720 -- nginx"
+    volumes:
+      - ${APP_DATA_DIR}/nginx/${BITCOIN_NETWORK}.conf:/etc/nginx/sites-enabled/dojo.conf
+    ports: 
+      - "$APP_SAMOURAI_DOJO_PORT:80"
+    expose:
+      - "80"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:123]: Will be removed.

File: apps/samourai-dojo/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      BITCOIND_RPC_PORT: $BITCOIN_RPC_PORT
+      BITCOIND_RPC_USER: $BITCOIN_RPC_USER
+      BITCOIND_RPC_PASSWORD: $BITCOIN_RPC_PASS
+      BITCOIND_ZMQ_RAWTXS: $BITCOIN_ZMQ_RAWTX_PORT
+      BITCOIND_ZMQ_BLK_HASH: $BITCOIN_ZMQ_HASHBLOCK_PORT
+      BITCOIND_SHUTDOWN_DELAY: 180
+      BITCOIND_EPHEMERAL_HS: "off"
+
+      # EXPLORER  
+      EXPLORER_INSTALL: "off"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:80]: EXPLORER_INSTALL needed here: https://code.samourai.io/dojo/samourai-dojo/-/blob/v1.8.1/docker/my-dojo/node/keys.index.js#L17

`BITCOIND_EPHEMERAL_HS` will be removed.

File: apps/samourai-dojo/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      NODE_ACTIVE_INDEXER: local_indexer
+      NODE_FEE_TYPE: ECONOMICAL
+
+      # SECURITY
+      NODE_API_KEY: "00000000000000000000000000000000"
+      NODE_ADMIN_KEY: "00000000000000000000000000000000"
+      NODE_JWT_SECRET: "00000000000000000000000000000000"
+
+      # BITCOIN
+      BITCOIND_INSTALL: "off"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:69]: Will be removed.

File: apps/samourai-dojo/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+      # EXPLORER  
+      EXPLORER_INSTALL: "off"
+
+      # INDEXER
+      INDEXER_INSTALL: "off"
+      INDEXER_IP: $ELECTRUM_IP
+      INDEXER_RPC_PORT: $ELECTRUM_PORT
+      INDEXER_BATCH_SUPPORT: inactive # 'active' for ElectrumX, 'inactive' otherwise
+    expose:
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:87]: Will be removed.

File: apps/samourai-dojo/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  nginx:
+    image: "louneskmt/dojo-nginx:1.5.0"
+    logging: *default-logging
+    restart: always
+    command: "/wait-for node:8080 --timeout=720 -- nginx"
+    volumes:
+      - ${APP_DATA_DIR}/nginx/${BITCOIN_NETWORK}.conf:/etc/nginx/sites-enabled/dojo.conf
+    ports: 
+      - "$APP_SAMOURAI_DOJO_PORT:80"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:121]: One would want to access the admin page from local network, don't you think? We can't expose Dojo Maintenance Tool only on Tor.

File: templates/bitcoin-sample.conf (hunk 2 above)
Comment (lukechilds) [reply templates/bitcoin-sample.conf:24]: Yes, perfect!

Comment (louneskmt) [reply templates/bitcoin-sample.conf:24]: See #468.

File: apps/samourai-dojo/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,127 @@
+version: "3.7"
+
+x-logging:
+  &default-logging
+  driver: journald
+  options:
+    tag: "umbrel-app {{.Name}}"
+
+services:
+  db:
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:10]: 8cd0827

File: apps/samourai-dojo/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        MYSQL_DATABASE: samourai-main
+        MYSQL_ROOT_PASSWORD: rootpassword
+        MYSQL_USER: samourai
+        MYSQL_PASSWORD: password
+    volumes:
+      - ${APP_DATA_DIR}/mysql:/var/lib/mysql
+    expose:
+      - "3306"
+  
+  node:
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:26]: 8cd0827

File: apps/samourai-dojo/docker-compose.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (lukechilds) [reply apps/samourai-dojo/docker-compose.yml:115]: Yeah potentially.

The thing is I don't imagine many people will actually use that, they'll just connect to Dojo and that's it.

So we should have a simple web page that shows Dojo connection details and Whirlpool connection details. That should be what's linked to (via clearnet or Tor) from the Umbrel dashboard.

Does the Dojo admin have the connection details listed? If so we could link directly to that, but then we don't have a good way to communicate the Whirlpool connection details.

File: apps/samourai-dojo/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+x-logging:
+  &default-logging
+  driver: journald
+  options:
+    tag: "umbrel-app {{.Name}}"
+
+services:
+  db:
+    image: "louneskmt/dojo-db:1.2.0"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml]: Added in b444272 (but needs some testing).

File: apps/samourai-dojo/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    ports:
+      - "$APP_SAMOURAI_DOJO_WHIRLPOOL_PORT:8898"
+    volumes:
+      - ${APP_DATA_DIR}/whirlpool:/home/whirlpool
+    networks:
+       default:
+         ipv4_address: $APP_SAMOURAI_DOJO_WHIRLPOOL_IP
+
+  nginx:
+    image: "louneskmt/dojo-nginx:1.5.0"
```
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml]: Added in d958774.

File: apps/samourai-dojo/docker-compose.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:120]: Let's first get it work and see what we can do from the admin page haha

File: templates/torrc-sample [hunk 15]
```diff
@@ -61,4 +61,12 @@ HiddenServicePort 80 <app-specter-desktop-ip>:<app-specter-desktop-port>
 HiddenServiceDir /var/lib/tor/app-btcpay-server
 HiddenServicePort 80 <app-btcpay-server-ip>:<app-btcpay-server-port>
 
+# samourai-dojo Hidden Service
+HiddenServiceDir /var/lib/tor/app-samourai-dojo
+HiddenServicePort 80 <app-samourai-dojo-ip>:<app-samourai-dojo-port>
```
Comment (lukechilds) [inline templates/torrc-sample:66]: Your pointing the Tor hidden service to port `<app-samourai-dojo-port>` (3005) when actually it's running on port 80 on the container and only mapped to port 3005 on the local network.

https://github.com/getumbrel/umbrel/pull/461/files#diff-fed5106248a13f4a034f9b50916d65709d37a2a8c62511565425891fefc38574R120

**Review by lukechilds — CHANGES_REQUESTED** [review]

File: templates/torrc-sample (hunk 15 above)
Comment (louneskmt) [reply templates/torrc-sample:66]: Good spot, that's why the hidden service didn't work.

Comment (louneskmt) [reply templates/torrc-sample]: Fixed in afdffbe.

File: apps/samourai-dojo/docker-compose.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (louneskmt) [reply apps/samourai-dojo/docker-compose.yml:123]: > So we should have a simple web page that shows Dojo connection details and Whirlpool connection details. That should be what's linked to (via clearnet or Tor) from the Umbrel dashboard.

Following our discussion on Keybase, we'll likely do as you said here, and keep Dojo admin page for power users maintenance.

File: apps/samourai-dojo/nginx/connect/index.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
+var baseRoute = "";
+
+if(bitcoinNetwork == "testnet") {
```
Comment (bguillaumat) [inline apps/samourai-dojo/nginx/connect/index.js:3]: Hey I think you've inversed your condition, no?

Comment (louneskmt) [reply apps/samourai-dojo/nginx/connect/index.js:3]: Yep 😁

File: apps/samourai-server/docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+      MYSQL_DATABASE: samourai-main
+      MYSQL_USER: samourai
+      MYSQL_PASSWORD: password
+
+      # NODEJS
+      NODE_GAP_EXTERNAL: 100
+      NODE_GAP_INTERNAL: 100
+      NODE_ADDR_FILTER_THRESHOLD: 1000
+      NODE_URL_OXT_API: https://api.oxt.me
+      NODE_URL_ESPLORA_API: https://blockstream.info/testnet
```
Comment (lukechilds) [inline apps/samourai-server/docker-compose.yml:53]: This is hardcoded to testnet.

Also do we even need these options set? Isn't this and `NODE_URL_OXT_API ` only used in the absence of an indexer like electrs?

Comment (louneskmt) [reply apps/samourai-server/docker-compose.yml:53]: > This is hardcoded to testnet.

That's normal, OXT is for main net and Explore for testnet.

But anyways, I just checked, we won't need any of these two as we use electrs al local indexer.

Comment (louneskmt) [reply apps/samourai-server/docker-compose.yml]: Resolved in fa2c985.

File: apps/samourai-server/nginx/connect/index.html [hunk 18]
```diff
@@ -82,6 +82,7 @@ <h2 class="text">How to connect Whirpool</h2>
             <li class="step">Click “Mix”.</li>
             <li class="step">Congratulations! Whirpool is now mixing your UTXOs on your Umbrel!</li>
         </ol>
+        <p><strong>Note:</strong> You'll need to open Whirlpool GUI and re-enter your password to continue mixing after restarting your Umbrel or doing an OTA update.</p>
```
Comment (mayankchhabra) [inline apps/samourai-server/nginx/connect/index.html:85]: Can we change `after restarting your Umbrel or doing an OTA update` to `after restarting or updating your Umbrel` to make sure the word "OTA" doesn't confuse users and there's no ambiguity around the update ("Umbrel update or Whirpool update?").

Comment (lukechilds) [reply apps/samourai-server/nginx/connect/index.html:85]: Good point

**Comment by AaronDewes** [comment]
Where is the source code for the whirlpool container?

**Comment by lukechilds** [comment]
@AaronDewes we're waiting on @louneskmt to push his fork but it's basically this with some minor modifications: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/whirlpool/Dockerfile

**Comment by AaronDewes** [comment]
> 
> 
> @AaronDewes we're waiting on @louneskmt to push his fork but it's basically this with some minor modifications: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/whirlpool/Dockerfile

Great! I'll try to implement whirlpools actual GUI next week if no one else does it until then

**Comment by louneskmt** [comment]
> @AaronDewes we're waiting on @louneskmt to push his fork but it's basically this with some minor modifications: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/whirlpool/Dockerfile

Here is my PR to Dojo source : https://code.samourai.io/dojo/samourai-dojo/-/merge_requests/196

Once it's merged (if it gets merged haha), the only manual change we will have to do is to switch from `uid:gid` `1110:1113` to `1000:1000` in Whirlpool `Dockerfile`. Unless we find another workaround which doesn't involve changing source.

**Comment by louneskmt** [comment]
> > 
> 
> > 
> 
> > @AaronDewes we're waiting on @louneskmt to push his fork but it's basically this with some minor modifications: https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/docker/my-dojo/whirlpool/Dockerfile
> 
> 
> 
> Great! I'll try to implement whirlpools actual GUI next week if no one else does it until then

Unfortunately that likely won't be possible, @lukechilds and I have already checked. Whirlpool GUI is an Electron app, with no clear separation between front end and backend.

**Comment by lukechilds** [comment]
It's definitely possible it's just a bit of a pain. All of the web ui <> node communication is done via electron ipc, so that all needs to be replaced with an HTTP API.

You could potentially even just stub out the electron ipc module with something that translates all the calls to HTTP.

However there might be some other issues too, e.g disable connection setup UI and pass all the local connection details directly in.

Definitely possible but a fairly big job. Feel free to have a crack at it though @AaronDewes, it would be a huge improvement to the Samourai experience on Umbrel if we could drop the Whirlpool GUI requirement!

**Comment by lukechilds** [comment]
Also the changes would need to be submitted upstream to Whirlpool GUI with either a runtime flag to toggle between Electron/web mode, or refactor the Electron app to do all communication via HTTP.

That way we can easily keep the Umbrel Whirlpool UI up to date with the official Whirlpool GUI since they would share the same codebase. What we really don't want to do is have our own separate web based fork of Whirlpool UI that we then need to continuously maintain to keep up to date.

It might be worth finding out if the Samourai devs would accept such a change before investing a lot of time into it.

**Comment by lukechilds** [comment]
@louneskmt are you able to push the exact source your containers are built from to GitHub?

Ideally then rebuild and push to Docker Hub so the checksums match. That way anyone is able to re-build from source to verify the images for themselves.

Once that's done, this is good to be merged!

**Comment by louneskmt** [comment]
@lukechilds As I've said, if this PR gets merged, it will be possible to build the images directly from the Dojo official source. 

Isn't it possible to find a workaround to fix this permissions issue (`uid:gid`) so it's possible to build Whirlpool image too without any modifications?

> Here is my PR to Dojo source : https://code.samourai.io/dojo/samourai-dojo/-/merge_requests/196
>
> Once it's merged (if it gets merged haha), the only manual change we will have to do is to switch from uid:gid 1110:1113 to 1000:1000 in Whirlpool Dockerfile. Unless we find another workaround which doesn't involve changing source.

**Comment by lukechilds** [comment]
Right but that requires manual changes, even just something tiny like changing the uids will result in a totally different checksum for the resulting Docker images so no one can verify the images unless they know the exact changes.

We should have code publicly available for all Docker images user in Umbrel that anyone can pull down and build to reproduce the images.

**Comment by lukechilds** [comment]
Right but that requires manual changes, even just something tiny like changing the uids will result in a totally different checksum for the resulting Docker images so no one can verify the images unless they know the exact changes.

We should have code publicly available for all Docker images user in Umbrel that anyone can pull down and build to reproduce the images.

**Comment by lukechilds** [comment]
>Isn't it possible to find a workaround to fix this permissions issue (uid:gid) so it's possible to build Whirlpool image too without any modifications?

Yes but we shouldn't do that. Umbrel should be in control of the container UID, not the Dockerfile. Otherwise a Dockerfile could run itself as root. For a Dockerfile to be compatible with Umbrel it needs to be able to be run under any UID.

**Comment by lukechilds** [comment]
@louneskmt if you could ping me once that's done and the latest images are pushed based on the public sources I'll add the checksums and do one last test (just in the unlikely case the latest images broke anything) and then we can merge!

**Comment by louneskmt** [comment]
> @louneskmt if you could ping me once that's done and the latest images are pushed based on the public sources I'll add the checksums and do one last test (just in the unlikely case the latest images broke anything) and then we can merge!

Samourai Dojo source code, edited for Umbrel, is now available at https://github.com/louneskmt/umbrel-samourai-dojo.

Latest version is tagged [here](https://github.com/louneskmt/umbrel-samourai-dojo/releases/tag/v1.8.1-umbrel).

Latest images are built and pushed to Docker Hub: [`dojo-nodejs`](https://hub.docker.com/repository/docker/louneskmt/dojo-nodejs) and [`dojo-whirlpool`](https://hub.docker.com/repository/docker/louneskmt/dojo-whirlpool).

Images tested and checksums added in e8ddbd2.

@lukechilds @mayankchhabra

**Review by lukechilds — APPROVED** [review]
Excellent stuff @louneskmt, really great job!
